File: src/main/java/com/harmonygames/harmonyengine/render/Texture.java
package com.harmonygames.harmonyengine.render;

import com.harmonygames.harmonyengine.Log;
import com.harmonygames.harmonyengine.math.Scale;
import org.lwjgl.BufferUtils;
import org.lwjgl.opengl.GL15;
import org.lwjgl.stb.STBImage;

import java.nio.ByteBuffer;
import java.nio.IntBuffer;

public class Texture {

    public static final int MIN_FILTER = GL15.GL_LINEAR;
    public static final int MAG_FILTER = GL15.GL_LINEAR;

    private String filepath;
    private int textureID;
    private int width, height;

    public Texture() {
        this.textureID = -1;
        this.width = -1;
        this.height = -1;
    }

    public Texture(String filepath, Scale scale) {
        this.filepath = filepath;

        // Generate texture on GPU
        this.textureID = GL15.glGenTextures();
        GL15.glBindTexture(GL15.GL_TEXTURE_2D, textureID);

        GL15.glTexParameteri(GL15.GL_TEXTURE_2D, GL15.GL_TEXTURE_MIN_FILTER, MIN_FILTER);
        GL15.glTexParameteri(GL15.GL_TEXTURE_2D, GL15.GL_TEXTURE_MAG_FILTER, MAG_FILTER);

        GL15.glTexImage2D(GL15.GL_TEXTURE_2D, 0, GL15.GL_RGB, width, height,
                0, GL15.GL_RGB, GL15.GL_UNSIGNED_BYTE, 0);
    }

    public void initialize() {
        // Generate Texture on GPU
        this.textureID = GL15.glGenTextures();
        GL15.glBindTexture(GL15.GL_TEXTURE_2D, textureID);

        GL15.glTexParameteri(GL15.GL_TEXTURE_2D, GL15.GL_TEXTURE_WRAP_S, GL15.GL_REPEAT);
        GL15.glTexParameteri(GL15.GL_TEXTURE_2D, GL15.GL_TEXTURE_WRAP_T, GL15.GL_REPEAT);

        GL15.glTexParameteri(GL15.GL_TEXTURE_2D, GL15.GL_TEXTURE_MIN_FILTER, MIN_FILTER);
        GL15.glTexParameteri(GL15.GL_TEXTURE_2D, GL15.GL_TEXTURE_MAG_FILTER, MAG_FILTER);

        IntBuffer width = BufferUtils.createIntBuffer(1);
        IntBuffer height = BufferUtils.createIntBuffer(1);
        IntBuffer channels = BufferUtils.createIntBuffer(1);

        STBImage.stbi_set_flip_vertically_on_load(true);

        ByteBuffer image = STBImage.stbi_load(filepath, width, height, channels, 0);

        if(image != null) {
            this.width = width.get(0);
            this.height = height.get(0);

            if(channels.get(0) == 3) { // RGB
                GL15.glTexImage2D(GL15.GL_TEXTURE_2D, 0, GL15.GL_RGB, width.get(0), height.get(0),
                        0, GL15.GL_RGB, GL15.GL_UNSIGNED_BYTE, image);
            } else if(channels.get(0) == 4) { // RGBA
                GL15.glTexImage2D(GL15.GL_TEXTURE_2D, 0, GL15.GL_RGBA, width.get(0), height.get(0),
                        0, GL15.GL_RGBA, GL15.GL_UNSIGNED_BYTE, image);
            } else {
                Log.fatalError("Unknown number of image channels '" + channels.get(0) + "' in image '" + filepath + "'");
            }

            STBImage.stbi_image_free(image);
        }
    }

    public void bind() {
        if(textureID == -1) Log.warn("Texture ID of image '" + filepath + "' holds a textureID pointer of -1!\n" +
                "\tThis may result in the texture not binding or an OpenGL error!!\n" +
                "\tThe textureID pointer should hold the memory location of the texture!");
        GL15.glBindTexture(GL15.GL_TEXTURE_2D, textureID);
    }

    public void unbind() {
        GL15.glBindTexture(GL15.GL_TEXTURE_2D, 0);
    }

    public int getTextureID() { return this.textureID; }
    public String getFilepath() { return this.filepath; }

    public int getWidth() { return this.width; }
    public int getHeight() { return this.height; }

    @Override
    public boolean equals(Object object) {
        if(!(object instanceof Texture)) return false;
        Texture texture = (Texture) object;
        return texture.getWidth() == this.width && texture.getHeight() == this.height && texture.getTextureID() == this.textureID
                && texture.getFilepath().equals(this.filepath);
    }
}

File: HarmonyEngine/src/Render/Texture.cpp
#include "Texture.h"

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#include "Renderer.h"

namespace HarmonyEngine {

// ====================================================================================
// Texture
// ====================================================================================

    void Texture::Initialize() {
        if (m_Filepath == nullptr) {
            Log::Warn("Can not create a texture without parameters!");
            return;
        }

        // Generate Texture on GPU
        glGenTextures(1, &m_TextureID);
        glBindTexture(GL_TEXTURE_2D, m_TextureID);

        // Setup Image Parameters
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, MIN_FILTER);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, MAG_FILTER);

        // Setup STB_Image
        stbi_set_flip_vertically_on_load(true);

        int channels;

        stbi_uc* image = stbi_load(m_Filepath, &m_Width, &m_Height, &channels, 0); // Load the image

        if (image != nullptr) {
            if (channels == 3) { // RGB Channels
                glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, m_Width, m_Height, 0, GL_RGB, GL_UNSIGNED_BYTE, image);
            } else if (channels == 4) { // RBGA Channels
                glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, m_Width, m_Height, 0, GL_RGBA, GL_UNSIGNED_BYTE, image);
            } else {
                Log::Error(std::string("Could not load image ") + m_Filepath +
                           std::string(", with unknown number of channels: ") + std::to_string(channels));
            }

            stbi_image_free(image);
        }

        Unbind();
    }

    void Texture::Initialize(const char* filepath, int width, int height) {
        m_Filepath = filepath;
        m_Width = width;
        m_Height = height;
        Texture::Initialize();
    }

    void Texture::Bind() {
        if (m_TextureID == -1) {
            Log::Warn("Texture ID of image '" + std::string(m_Filepath) + "' holds a textureID pointer of -1!\n" +
                      "\tThis may result in the texture not binding or an OpenGL error!!\n" +
                      "\tThe textureID pointer should hold the memory location of the texture!");
        }

        glBindTexture(GL_TEXTURE_2D, m_TextureID);
    }

    void Texture::Unbind() {
        glBindTexture(GL_TEXTURE_2D, 0);
    }

// ====================================================================================
// SpriteSheet
// ====================================================================================
    void SpriteSheet::AssignToSprite(Quad* quad, const int& spriteX, const int& spriteY) {
        float normalizedX = (float) spriteX * m_NormalizedSpriteWidth;
        float normalizedY = (float) spriteY * m_NormalizedSpriteHeight;

        quad->V0.TextureCoord = {normalizedX, normalizedY};
        quad->V1.TextureCoord = {normalizedX, normalizedY + m_NormalizedSpriteHeight};
        quad->V2.TextureCoord = {normalizedX + m_NormalizedSpriteWidth, normalizedY + m_NormalizedSpriteHeight};
        quad->V3.TextureCoord = {normalizedX + m_NormalizedSpriteWidth, normalizedY};
    }

}